cpp/locality/Voronoi.cc

```diff
@@ -0,0 +1,195 @@
+                    if (ridge_vertices[ridge_vert_id] == -1) {
+                        add_valid_bonds(local_bonds, i, expanded_ids[i], j, expanded_ids[j], N, 0);
+                        weighted_bond = false;
+                        break;
+                    } else {
+                        current_ridge_vertex_ids.push_back(ridge_vertices[ridge_vert_id]);
+                    }
+                }
+
+                if (weighted_bond) {
+                    if (box.is2D()) {
+                        // 2D weight is the length of the ridge edge
+                        int v1ind = current_ridge_vertex_ids[0];
+                        int v2ind = current_ridge_vertex_ids[1];
+                        vec3<double> v1 = vertices[v1ind];
+                        vec3<double> v2 = vertices[v2ind];
+                        vec3<float> rij(box.wrap(v1 - v2));
+                        weight = sqrt(dot(rij, rij));
+                    } else {
+                        // 3D weight is the area of the ridge facet
+                        // Create a vector of all vertices for this facet
+                        vector< vec3<double> > vertex_coords;
+
+                        for (auto ridge_vert_id = current_ridge_vertex_ids.begin();
+                            ridge_vert_id != current_ridge_vertex_ids.end();
+                            ++ridge_vert_id)
+                        {
+                            vec3<double> vert = vertices[*ridge_vert_id];
+                            vertex_coords.push_back(vert);
+                        }
+
+                        // Code below is adapted from http://geomalgorithms.com/a01-_area.html
+                        // Get a unit normal vector to the polygonal facet
+                        // Every facet has at least 3 vertices
+                        vec3<double> r01(vertex_coords[1] - vertex_coords[0]);
+                        vec3<double> r12(vertex_coords[2] - vertex_coords[1]);
+                        vec3<double> norm_vec = cross(r01, r12);
+                        norm_vec /= sqrt(dot(norm_vec, norm_vec));
+
+                        // Determine projection axis (x=0, y=1, z=2)
+                        double c0_component = std::max(std::max(abs(norm_vec.x), abs(norm_vec.y)), abs(norm_vec.z));
+                        int c0 = 2;
+                        if (c0_component == abs(norm_vec.x)) {
+                            c0 = 0;
+                        } else if (c0_component == abs(norm_vec.y)) {
+                            c0 = 1;
+                        }
+
+                        double projected_area = 0;
+                        int n_verts = vertex_coords.size();
+                        for (int step = 0; step < n_verts; step++)
+                        {
+                            int n1 = step % n_verts;
+                            int n2 = (step + 1) % n_verts;
+                            int n3 = (step - 1 + n_verts) % n_verts;
```

Comment:
Is there any reason you can't use unsigned ints for all of these (including `step`)? If there are compiler warnings associated with the `step-1` being potentially negative, you might be able to trick it by rewriting this as ((step + n_verts) - 1) or something like that. If not, then leave it.